### PR TITLE
✨ Add PyTorch image classification example

### PR DESCRIPTION
**Pull request by nateraw**
# What does this PR do?

Adds PyTorch image classification example. For now, it uses `torchvision.datasets.ImageFolder` to load local image folders (just like the flax image classification example). In the future, we will switch to using the `datasets` package's image folder (once it exists). 

Marking as draft for now as I'm still working through cleaning up changes I made from [this example](https://github.com/nateraw/vision/tree/main/image-classification) I wrote earlier that uses `datasets` instead.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by NielsRogge** [comment]
Nice!! Relevant for #13080

File: examples/pytorch/vision/run_image_classification.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+                "Use --overwrite_output_dir to overcome."
+            )
+        elif last_checkpoint is not None and training_args.resume_from_checkpoint is None:
+            logger.info(
+                f"Checkpoint detected, resuming training at {last_checkpoint}. To avoid this behavior, change "
+                "the `--output_dir` or add `--overwrite_output_dir` to train from scratch."
+            )
+
+    normalize = T.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5])
+    train_dataset = ImageFolder(
```
Comment (sgugger) [inline examples/pytorch/vision/run_image_classification.py:172]: It would be nice to use the datasets library here, or support both use cases (like run_glue for text classification).

File: examples/pytorch/vision/run_image_classification.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            data_args.validation_dir,
+            T.Compose(
+                [
+                    T.Resize(data_args.image_size),
+                    T.CenterCrop(data_args.image_size),
+                    T.ToTensor(),
+                    normalize,
+                ]
+            ),
+        )
```
Comment (sgugger) [inline examples/pytorch/vision/run_image_classification.py:200]: Let's define the validation transform in another line as black makes this hard to read.

File: examples/pytorch/vision/run_image_classification.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    train_dataset = ImageFolder(
+        data_args.train_dir,
+        T.Compose(
+            [
+                T.RandomResizedCrop(data_args.image_size),
+                T.RandomHorizontalFlip(),
+                T.ToTensor(),
+                normalize,
+            ]
+        ),
```
Comment (sgugger) [inline examples/pytorch/vision/run_image_classification.py:181]: Let's define the training transform on another line as black makes this hard to read.

File: examples/pytorch/vision/run_image_classification.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+                    T.Resize(data_args.image_size),
+                    T.CenterCrop(data_args.image_size),
+                    T.ToTensor(),
+                    normalize,
+                ]
+            ),
+        )
+        if data_args.test_dir is not None
+        else None
+    )
```
Comment (sgugger) [inline examples/pytorch/vision/run_image_classification.py:221]: Let's reuse the validation transform from before and do an if then else on several lines. There is no point using the one-line syntax `foo = bar if xxx else yyy` when black splits it in several lines.

File: examples/pytorch/vision/run_image_classification.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        indices = torch.randperm(data_args.max_train_samples).tolist()
+        train_dataset = torch.utils.data.Subset(train_dataset, indices)
+    if data_args.max_eval_samples is not None and data_args.max_eval_samples < len(eval_dataset):
+        indices = torch.randperm(data_args.max_eval_samples).tolist()
+        eval_dataset = torch.utils.data.Subset(eval_dataset, indices)
+    if data_args.max_predict_samples is not None and data_args.max_predict_samples < len(test_dataset):
+        indices = torch.randperm(data_args.max_predict_samples).tolist()
+        test_dataset = torch.utils.data.Subset(test_dataset, indices)
+
+    # NOTE - We aren't directly using this feature extractor since we defined custom transforms above.
```
Comment (sgugger) [inline examples/pytorch/vision/run_image_classification.py:259]: Explain a bit more why we are defining it anyway? It doesn't make sense otherwise.

**Review by sgugger — APPROVED** [review]
Thanks a lot for adding this! I left a few comments.

File: examples/pytorch/vision/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+-f https://download.pytorch.org/whl/torch_stable.html
+torch==1.9.0+cpu 
+-f https://download.pytorch.org/whl/torch_stable.html
+torchvision==0.10.0+cpu
```
[the remark this answers is not in the document]
Comment (nateraw) [reply examples/pytorch/vision/requirements.txt]: resolved!

File: examples/pytorch/vision/run_image_classification.py (hunk 1 above)
Comment (nateraw) [reply examples/pytorch/vision/run_image_classification.py]: Using `datasets` now. For image folder, we can use [`nateraw/image-folder`](https://huggingface.co/datasets/nateraw/image-folder/) until the [official implementation](https://github.com/huggingface/datasets/pull/2830) gets merged.

File: examples/pytorch/vision/run_image_classification.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+import math
+import os
+import sys
+from dataclasses import dataclass, field
+from typing import Optional
+
+import datasets
+import numpy as np
+import torch
+import torchvision.transforms as T
```
[the remark this answers is not in the document]
Comment (nateraw) [reply examples/pytorch/vision/run_image_classification.py]: importing them directly now

**Comment by NielsRogge** [comment]
I'll review this PR in detail (thanks for working on this!). Regarding the fixtures for the tests, I've recently moved these files to the hf-internal-testing organization on the [hub](https://huggingface.co/hf-internal-testing). This makes it more clear, as otherwise these fixture files are also downloaded when people do a `git clone` of the library.

File: examples/pytorch/test_examples.py [hunk 8]
```diff
@@ -340,3 +342,35 @@ def test_run_translation(self):
...
+    def test_run_image_classification(self):
+        stream_handler = logging.StreamHandler(sys.stdout)
+        logger.addHandler(stream_handler)
+
+        tmp_dir = self.get_auto_remove_tmp_dir()
+        testargs = f"""
+            run_image_classification.py
+            --output_dir {tmp_dir}
+            --model_name_or_path google/vit-base-patch16-224-in21k
+            --train_dir tests/fixtures/tests_samples/cats_and_dogs/
```
Comment (NielsRogge) [inline examples/pytorch/test_examples.py:355]: Can we add the cats and dogs as a dataset to the hub under the [hf-internal-testing](https://huggingface.co/hf-internal-testing) organization?

File: examples/pytorch/vision/README.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Image classification examples
+
+The following example showcases how to fine-tune a `ViT` for image-classification using PyTorch.
```
Comment (NielsRogge) [inline examples/pytorch/vision/README.md:19]:
```suggestion
The following examples showcase how to fine-tune a `ViT` for image-classification using PyTorch.
```

File: examples/pytorch/vision/README.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+## Sharing your model on 🤗 Hub
+
+0. If you haven't already, [sign up](https://huggingface.co/join) for a 🤗 account
+
+1. Make sure you have `git-lfs` installed.
+
+```bash
+$ apt install git-lfs
+```
```
Comment (NielsRogge) [inline examples/pytorch/vision/README.md:115]:
```suggestion
1. Make sure you have `git-lfs` installed and git setup.

```bash
$ apt install git-lfs
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
```
```
Not sure if this is needed, but if you run the script in Google Colab for example then you also need to setup git (based on Sylvain's notebooks such as [this one](https://github.com/huggingface/notebooks/blob/master/examples/text_classification.ipynb)).

File: examples/pytorch/vision/README.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+2. Log in using `huggingface-cli`
```
Comment (NielsRogge) [inline examples/pytorch/vision/README.md:117]:
```suggestion
2. Log in with your HuggingFace account credentials using `huggingface-cli`
```

File: examples/pytorch/vision/README.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+$ huggingface-cli login
+# ...follow the prompts
+```
+
+3. When running the script, pass the following arguments:
+
+```bash
+python run_image_classification.py \
+    --push_to_hub \
+    --push_to_hub_model_id <name-your-model> \
```
Comment (NielsRogge) [inline examples/pytorch/vision/README.md:129]: Is it also possible to push to a certain organization?

File: examples/pytorch/vision/run_image_classification.py [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+    HfArgumentParser,
+    Trainer,
+    TrainingArguments,
+)
+from transformers.trainer_utils import get_last_checkpoint
+from transformers.utils import check_min_version
+from transformers.utils.versions import require_version
+
+
+""" Finetuning a 🤗 Transformers model for image classification"""
```
Comment (NielsRogge) [inline examples/pytorch/vision/run_image_classification.py:52]:
```suggestion
""" Fine-tuning a 🤗 Transformers model for image classification"""
```

File: examples/pytorch/vision/run_image_classification.py [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+        return im.convert("RGB")
+
+
+@dataclass
+class DataTrainingArguments:
+    """
+    Arguments pertaining to what data we are going to input our model for training and eval.
+    Using `HfArgumentParser` we can turn this class
+    into argparse arguments to be able to specify them on
+    the command line.
```
Comment (NielsRogge) [inline examples/pytorch/vision/run_image_classification.py:77]:
```suggestion
    Using `HfArgumentParser` we can turn this class into argparse arguments to be able to 
    specify them on the command line.
```

File: examples/pytorch/vision/run_image_classification.py [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+        },
+    )
+    max_eval_samples: Optional[int] = field(
+        default=None,
+        metadata={
+            "help": "For debugging purposes or quicker training, truncate the number of evaluation examples to this "
+            "value if set."
+        },
+    )
+    image_size: Optional[int] = field(default=224, metadata={"help": " The size (resolution) of each image."})
```
Comment (NielsRogge) [inline examples/pytorch/vision/run_image_classification.py:105]: Will images be squared? Or is it just the smaller edge of the image that will be matched to this number?

File: examples/pytorch/vision/run_image_classification.py [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+        default="nateraw/image-folder", metadata={"help": "Name of a dataset from the datasets package"}
+    )
+    dataset_config_name: Optional[str] = field(
+        default=None, metadata={"help": "The configuration name of the dataset to use (via the datasets library)."}
+    )
+    train_dir: Optional[str] = field(default=None, metadata={"help": "A folder containing the training data."})
+    validation_dir: Optional[str] = field(default=None, metadata={"help": "A folder containing the validation data."})
+    train_val_split: Optional[float] = field(
+        default=0.15, metadata={"help": "Percent to split off of train for validation."}
+    )
```
Comment (NielsRogge) [inline examples/pytorch/vision/run_image_classification.py:90]: Hmm so you have a training dataset, a validation dataset and a test set. The `validation_dir` is actually the test set? And the `train_val_split` will split the training into training + validation?

**Review by NielsRogge — COMMENTED** [review]
Overall LGTM! Just some small comments.

Most importantly, it would be great to remove the fixtures files from this PR in favor of a HuggingFace dataset.

**Comment by sgugger** [comment]
Last nit on my side: can we move the vision folder to be `image-classification`? We will have other kinds of vision examples in the future.

File: examples/pytorch/vision/README.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
+<!---
```
Comment (patrickvonplaten) [inline examples/pytorch/vision/README.md:1]: @sgugger - is `pytorch/vision` not a bit too general? Given that we have `examples/pytorch/text-classification/run_glue.py` I think the better structure would be `examples/pytorch/image-classification/run_image_classification.py` here no?

Comment (patrickvonplaten) [reply examples/pytorch/vision/README.md:1]: Never mind - just saw your comment!

File: examples/pytorch/vision/run_image_classification.py (hunk 15 above)
Comment (nateraw) [reply examples/pytorch/vision/run_image_classification.py:105]: You're correct, this value is passed directly to torchvision's `Resize`

File: examples/pytorch/vision/README.md (hunk 12 above)
Comment (nateraw) [reply examples/pytorch/vision/README.md:129]: Yup, this is all handled through trainer arguments

File: examples/pytorch/vision/run_image_classification.py (hunk 16 above)
Comment (nateraw) [reply examples/pytorch/vision/run_image_classification.py]: there's no test set here. There's just train + validation. If 'validation' key is not found in dataset, we create a split off of train for validation and set it to the 'validation' key of the dataset dict.

File: examples/pytorch/test_examples.py (hunk 8 above)
Comment (nateraw) [reply examples/pytorch/test_examples.py:355]: We can, but this is actually testing the fact that the script works on local image folders. I can issue another PR that directly pulls them down into an imagefolder-like cache dir to test. But I'll leave this as-is for now if its not a big deal.

**Comment by nateraw** [comment]
Ok, addressed most of the comments. Merging as-is for now.

@NielsRogge I did not address these two items, however I can in future PRs (if need be):

- Adding test data to `datasets` library. 
- Adjusting train/validation/test split logic.

**Comment by Hecim1984** [comment]
13134
